File: leetcode/go/1723.find-minimum-time-to-finish-all-jobs.go
import (
    "sort"
)
func minimumTimeRequired(jobs []int, k int) int {
    sort.Sort(sort.Reverse(sort.IntSlice(jobs)))
    ans := 0
    for _, j := range jobs {
        ans += j
    }
    workers := make([]int, k)
    backtrack(jobs, 0, workers, &ans)
    return ans
}

func backtrack(jobs []int, job_id int, workers []int, ans *int) {
    if job_id == len(jobs) {
        mx := workers[0]
        for i:=1; i<len(workers); i++ {
            mx = max(mx, workers[i])
        }
        *ans = min(*ans, mx)
        return
    }
    for worker_id:=0; worker_id < len(workers); worker_id++ {
        if workers[worker_id] + jobs[job_id] >= *ans {
            continue
        }
        workers[worker_id] += jobs[job_id]
        backtrack(jobs, job_id + 1, workers, ans)
        workers[worker_id] -= jobs[job_id]
        if workers[worker_id] == 0 {
            return
        }
    }
}

File: leetcode/rust-leetcode/src/1723.find-minimum-time-to-finish-all-jobs.rs
impl Solution {
    pub fn minimum_time_required(mut jobs: Vec<i32>, k: i32) -> i32 {
        let k = k as usize;
        jobs.sort();
        jobs.reverse();
        fn backtrack(jobs: &Vec<i32>, job_id: usize, workers: &mut Vec<i32>, ans: &mut i32, bitmask: u32) {
            if job_id == jobs.len() {
                *ans = (*ans).min(*workers.iter().max().unwrap());
                return;
            }
            for worker_id in 0..workers.len() {
                workers[worker_id] += jobs[job_id];
                if workers[worker_id] < *ans {
                    backtrack(jobs, job_id+1, workers, ans, bitmask);
                }
                workers[worker_id] -= jobs[job_id];
                if workers[worker_id] == 0 {
                    return;
                }
            }
        }
        let mut ans = jobs.iter().sum::<i32>();
        let mut workers = vec![0;k];
        backtrack(&jobs, 0, &mut workers, &mut ans, 0);
        ans
    }
}


